db/migrate/20190906165501_remove_broken_variants_from_carts.rb

```diff
@@ -0,0 +1,14 @@
+class RemoveBrokenVariantsFromCarts < ActiveRecord::Migration
+  def up
+    # Removes line_items from open carts where the variant has a hard-deleted price
+
+    variants_with_prices = Spree::Variant.
+      joins('INNER JOIN spree_prices ON (spree_variants.id = spree_prices.variant_id)').
+      select('spree_variants.id')
+
+    Spree::LineItem.
+      joins(:variant).where('spree_variants.id NOT IN (?)', variants_with_prices).
+      joins(:order).merge(Spree::Order.where(state: 'cart')).
+      delete_all
```

Comment:
1. is this the only migration we need here? I cant remember but maybe there is more data that needs to be cleaned up?

2. I think you should invert the first query because variants_with_prices can be too big
- I think you need to use arel for this (I never used it)
variants_without_price = select * from spree_variants v where not exists (select 1 from spree_prices where variant_id = v.id)
and then same second query with IN instead of NOT IN

Reply:
I've updated the queries and tested them more thoroughly in the Rails console. :+1:

I'm not sure how much more we can clear up in terms of data. We can't restore hard-deleted prices, and I don't think we can reasonably remove line_items from completed orders.